workbench-core/authorization/src/dynamicAuthorization/dynamicAuthorizationService.ts

```diff
@@ -0,0 +1,203 @@
+
+/**
+ * Request object for DynamicAuthorizationService's isAuthorizedOnSubjectRequest
+ */
+export interface IsAuthorizedOnSubjectRequest {
+  /**
+   * {@link DynamicOperation}
+   */
+  dynamicOperation: DynamicOperation;
+}
+
+export class DynamicAuthorizationService {
+  private _dynamicPermissisonsPlugin: DynamicPermissionsPlugin;
+  private _authorizationPlugin: AuthorizationPlugin;
+
+  public constructor(params: {
+    dynamicPermissionsPlugin: DynamicPermissionsPlugin;
+    authorizationPlugin: AuthorizationPlugin;
+  }) {
+    this._dynamicPermissisonsPlugin = params.dynamicPermissionsPlugin;
+    this._authorizationPlugin = params.authorizationPlugin;
+  }
+
+  /**
+   * Checks whether a {@link AuthenticatedUser} is authorized to perform {@link isAuthorizedOnSubjectRequest}
+   * @param user - {@link AuthenticatedUser}
+   * @param isAuthorizedOnSubjectRequest - {@link IsAuthorizedOnSubjectRequest}
+   *
+   * @throws - {@link ForbiddenError} when {@link AuthenticatedUser} is not authorized.
+   */
+  public async isAuthorizedOnSubject(
+    user: AuthenticatedUser,
+    isAuthorizedOnSubjectRequest: IsAuthorizedOnSubjectRequest
+  ): Promise<void> {
+    const { id, roles } = user;
+    const { subjectId, subjectType, action } = isAuthorizedOnSubjectRequest.dynamicOperation;
+    const { groupIds } = await this._dynamicPermissisonsPlugin.getUserGroups({
+      userId: id
+    });
+    const groupIdIdentities = Array.from(new Set<string>([...groupIds, ...roles])).map(
+      (groupId): Identity => {
+        return {
+          identityType: 'GROUP',
+          identityId: groupId
+        };
+      }
+    );
+    const userIdentity: Identity = {
+      identityType: 'USER',
+      identityId: id
+    };
+    const identities: Identity[] = [userIdentity, ...groupIdIdentities];
+    //Check for subject id permissions
+    const { identityPermissions } = await this._dynamicPermissisonsPlugin.getIdentityPermissionsBySubject({
+      subjectType,
+      subjectId,
+      action,
+      identities
+    });
+    //Check for wildcard permissions
+    const { identityPermissions: wildCardIdentityPermissions } =
+      await this._dynamicPermissisonsPlugin.getIdentityPermissionsBySubject({
+        subjectType,
+        subjectId: '*',
+        action,
+        identities
+      });
+
+    const allUserPermissions: IdentityPermission[] = [...identityPermissions, ...wildCardIdentityPermissions];
+
+    await this._authorizationPlugin.isAuthorizedOnDynamicOperations(allUserPermissions, [
+      isAuthorizedOnSubjectRequest.dynamicOperation
+    ]);
+  }
+  /**
+   * Create an authorization group
+   * @param createGroupRequest - {@link CreateGroupRequest}
+   *
+   * @returns - {@link CreateGroupResponse}
+   *
+   * @throws - {@link GroupAlreadyExistsError} Can not create a group that already exists
+   */
+  public async createGroup(createGroupRequest: CreateGroupRequest): Promise<CreateGroupResponse> {
```

Comment:
Why are we creating pass through methods instead of making the plugin public and accessing the plugin directly?

Reply:
This could be done at the consumer's discretion, but it is rather easier to encapsulate this under one service

Reply:
I think some sort of concept is missing here: the functions on the `DynamicAuthorizationService` and `DynamicAuthorizationService` are almost identical, but they're independent of each other. It looks like there is maybe an interface that both of these objects conform to at a minimum. Maybe there should be two separate classes, because it seems like we've joined Auth check functionality with group management.

Reply:
What would be the benefit of separation of concerns for authz check and group management, given they both currently live under the authorization service?

Reply:
In this specific case, we'd be improving adherence to the Single Responsibility Principle. If I understand correctly, these two functionalities are not  actually related to each other from a responsibility perspective: one checks for authorization, the other is a CRUD interface for Group management.

Are you going for a single class that the user can call to manage all things Auth related? If so, I don't think that is the direction we'd want to head because as we add more functionality the interface for that service gets more and more complex as the class takes on multiple responsibilities.